torchaudio/prototype/datasets/bucketize_sampler.py

```diff
@@ -0,0 +1,141 @@
+
+        Returns:
+            (dict[int, Tensor]): A dictionary in which the key is the bucket index, the value is
+                the Tensor of corresponding sample indices.
+        """
+        buckets = {}
+
+        boundaries = [min_len - 1]
+        interval = (max_len - min_len) // num_buckets
+        for i in range(1, num_buckets):
+            boundaries.append(min_len + i * interval)
+        boundaries.append(max_len + 1)
+        bucket_ids = torch.bucketize(torch.tensor(lengths), torch.tensor(boundaries))
+        for i in indices:
+            bucket_id = bucket_ids[self.index2iter[i]]
+            if bucket_id in buckets:
+                buckets[bucket_id].append(i)
+            else:
+                buckets[bucket_id] = [i]
+        for k in buckets:
+            buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+            if self.shuffle:
+                buckets[k] = buckets[k][torch.randperm(buckets[k].size(0))]
+        return buckets
+
+    def __iter__(self) -> Iterator[List[int]]:
+        iter_list = []
+        total_len = 0
+        batch = []
+        if self.max_token_count is not None:
+            for k in self.buckets.keys():
+                for i in range(self.buckets[k].size(0)):
+                    index = self.buckets[k][i]
+                    if len(batch) == 0 and self.lengths[self.index2iter[int(index)]] > self.max_token_count:
+                        raise ValueError(
+                            f"The length of the sample is {self.lengths[self.index2iter[int(index)]]}, "
+                            f"which is greater than ``max_token_count``: {self.max_token_count}."
+                            "Make sure you set the correct ``max_len`` and ``max_token_count``."
+                        )
+                    elif total_len + self.lengths[self.index2iter[int(index)]] <= self.max_token_count:
+                        batch.append(int(index))
+                        total_len += self.lengths[self.index2iter[int(index)]]
+                    else:
+                        iter_list.append(batch)
+                        batch = []
+                        total_len = 0
+
+            if len(batch) > 0:
+                iter_list.append(batch)
+        else:
+            for k in self.buckets.keys():
+                for i in range(self.buckets[k].size(0)):
+                    index = self.buckets[k][i]
+                    if total_len == self.batch_size:
+                        iter_list.append(batch)
+                        batch = [index]
+                        total_len = 1
+                    else:
+                        batch.append(index)
+                        total_len += 1
+
+        for batch in iter_list:
+            yield batch
+
+    def __len__(self):
+        if self.batch_size:
+            return len(self.lengths) // self.batch_size
+        else:
+            return sum(self.lengths) // self.max_token_count
```

Comment:
this doesn't hold in general

for example, if the underlying dataset comprises 10 samples of length 6 each, and `max_token_count` is 10, `__len__` would return 6, but the actual number of batches returned would be 10

Reply:
Yep, but I don't know how to estimate the correct number of batches, especially when ``shuffle`` is True.

Reply:
yeah i don't think it's possible. ideally this shouldn't return a numerical value to reflect that. what do callers expect?

Reply:
Ideally for each epoch they expect to iterate all the data once. So the lengths should reflect the number of batches in each iteration.
Is it possible to update the ``__len__`` of the sampler in the ``__iter__`` method?

Reply:
hmm i'd rather not, since the length isn't guaranteed to be stable and we have no control over the order in which the user calls `__iter__` and `__len__`. `__len__` should simply be invalidated somehow in cases where we're batching by token count and shuffling